packages/design-system/src/components/primitives/dnd/use-drag.ts

```diff
@@ -75,67 +73,64 @@ export const useDrag = <DragItemData>({
         return false;
       }
 
-      dragItemData.current = data;
-
+      state.current.dragItemData = data;
       return true;
     },
-    onMoveStart({
-      clientX: x,
-      clientY: y,
-      target,
-    }: {
-      clientX: number;
-      clientY: number;
-      target: HTMLElement;
-    }) {
+    onMoveStart({ clientX: x, clientY: y }) {
       state.current = {
+        ...state.current,
         status: "pending",
-        x,
-        y,
         initialX: x,
         initialY: y,
-        shifts: 0,
       };
-      if (dragItemData.current !== undefined) {
-        onStart({ target, data: dragItemData.current });
-      }
     },
-    onMove({ clientX: x, clientY: y }: { clientX: number; clientY: number }) {
+    onMove({ clientX: x, clientY: y }) {
       // We want to start dragging only when the user has moved more than startDistanceThreashold.
       if (
         state.current.status === "pending" &&
-        Math.abs(x - state.current.x) < startDistanceThreashold &&
-        Math.abs(y - state.current.y) < startDistanceThreashold
+        Math.abs(x - state.current.initialX) < startDistanceThreashold &&
+        Math.abs(y - state.current.initialY) < startDistanceThreashold
       ) {
         return;
       }
 
-      // We shouldn't be in non-pending state when we get here, unles user has ended the drag and yet somehow we ended up with onMove() call.
       if (
-        state.current.status === "pending" ||
-        state.current.status === "dragging"
+        state.current.status === "pending" &&
+        state.current.dragItemData !== undefined
       ) {
-        state.current = {
-          status: "dragging",
-          x,
-          y,
-          initialX: state.current.initialX,
-          initialY: state.current.initialY,
-          shifts: state.current.shifts,
-        };
+        onStart({ data: state.current.dragItemData });
+        state.current.status = "dragging";
       }
 
-      onMove({ x, y });
-      detectShift();
+      if (state.current.status === "dragging") {
+        onMove({ x, y });
+        detectShift(x);
+      }
     },
-    onMoveEnd(event) {
-      state.current = initialState;
-      onEnd({ isCanceled: event === "canceled" });
+    onMoveEnd({ isCanceled }) {
+      // A drag is basically a very slow click.
+      // But we don't want it to register as a click.
+      if (isCanceled === false && state.current.status === "dragging") {
+        const handleClick = (event: MouseEvent) => {
+          event.preventDefault();
+          event.stopPropagation();
+        };
+        window.addEventListener("click", handleClick, true);
+        setTimeout(() => {
```

Comment:
why not removing the listener inside the handleClick instead? there must be a more deterministic way of handling it

Reply:
I did that at first, but I'm not 100% sure a click event will always occur. If it doesn't, a random future click will be canceled. I think `setTimeout` is more robust in this case even though a bit hacky.

Reply:
yeah, though its very indeterministic with javascript when this setTimeout is going to run, what if js runtime decides to run it in 1ms after you called it and the click actually occurs after that for some reason?

Reply:
it can also cause issues with e2e testing in the future

Reply:
I think the actual problem here is that the event that handles pointer/mouse up doesn't preventDefault where it should, so you are now upstream trying to workaround the situation

Reply:
I thought that too. But I tried adding it to our `pointerup` and `pointerdown` handlers, and it didn't work

Reply:
The case when `click` might not happen that I was thinking of is when you release the button outside of browser window / iframe. Need to check what happens in that case.

If it still generates a click, then it's probably ok to unsubscribe in the event handler as I'm not aware of other cases.

If it doesn't, then we probably need to both unsubscribe in the handler and after something like 300ms as a fallback.

Reply:
Actually there's a better option: unsubscribe in the handler, prevent default only if less than 300ms have passed since we subscribed.

Reply:
timeout based logic is flawed in any case, since we know when drag interaction started, we can bind there an event like mouseup and prevent a click there

Reply:
jfyi click is mousedown+mouseup, so you should be able to prevent click at both mousedown and mouseup, haven't tested mouseup, just assuming it works based on the mental model I have

Reply:
Not sure it's because of pointer event, here they do similar thing with `mouseup`: https://stackoverflow.com/a/20290312/478603

But I will try

Reply:
`setTimeout` removed.

Preventing default in `mouseup` doesn't prevent the click.
Click does happen even when releasing outside of browser window, but I still added a check for 300ms just in case.